### PR TITLE
add pytorch datasets and dataloaders docs

### LINKED ISSUE #4933: [Concept Entry] PyTorch Datasets and Dataloaders
### Reason/inspiration (optional)

We would like a new entry on the `Datasets and Dataloaders` concept under PyTorch. The entry should go in a new file under `docs/content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md`.

The entry should include:

- An introduction to the concept
- A `Syntax` section that provides the syntax for the concept (if applicable)
- An `Example` section that provides an example demonstrating the concept in use (if applicable)

Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entrty for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by wejk-ewjslkj**
### Description

I have created a new entry titled **"PyTorch Datasets and DataLoaders"**. This entry provides an overview of the `Dataset` and `DataLoader` abstractions in PyTorch, including instructions on how to create a custom dataset class and utilize the `DataLoader` for efficient data handling. The entry also includes examples of using built-in datasets and configuring the `DataLoader`. The motivation for this contribution is to help users understand and effectively use these fundamental PyTorch tools in their machine learning workflows.

### Issue Solved

Closes #4933

### Type of Change

- [x] Adding a new entry
- [ ] Editing an existing entry (fixing a typo, bug, issues, etc)
- [ ] Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+## Creating a Custom Dataset
+
+To create a custom dataset in PyTorch, subclass `torch.utils.data.Dataset` and override the following methods:
+
+- `__init__(self)`: Initializes the dataset, typically loading data into memory or setting up file paths.
+- `__len__(self)`: Returns the total number of samples in the dataset.
+- `__getitem__(self, idx)`: Retrieves a sample and its corresponding label at the specified index `idx`.
+
+### Example: Custom Dataset Class
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:26]:
``` suggestion
### Example: Custom Dataset Class

```
Instead of mentioning `custom dataset classes` in front of the example, please note it in the description related to the example block.

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        self.labels = labels
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        sample = self.data[idx]
+        label = self.labels[idx]
+        return sample, label
+```
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:44]:
``` suggestion
```py
import torch
from torch. utils.data import Dataset

class CustomDataset(Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels

    def __len__(self):
        return len(self.data)

    def __getitem__(self, IDX):
        sample = self.data[IDX]
        label = self.labels[IDX]
        return sample, label
```

```
The supported code block for `python` is ```py```, so the code should be written inside ``` code block instead of writing it inside ```python``` code block.

In the above code, there is a small typo mistake, the parameter `IDX` is defined in the `getitem` method, but when you're referencing it inside the method you are referring to it as `idx` which is not correct. We should use the same variable name inside the method that we have used at the time of the variable declaration.

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+## DataLoader
+
+The `DataLoader` class in PyTorch is used to load data in batches, which is essential for training neural networks efficiently. It also provides options for shuffling the data and loading it in parallel using multiprocessing.
+
+### Example: Using DataLoader
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:50]:
```suggestion
### Example

```
Instead of mentioning ` Using DataLoader` in front of the example header, please note it in the description related to the example block.

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+```python
+from torch.utils.data import DataLoader
+
+# Assuming custom_dataset is an instance of CustomDataset
+data_loader = DataLoader(dataset=custom_dataset, batch_size=4, shuffle=True)
+
+for batch in data_loader:
+    samples, labels = batch
+    # Training code here
+```
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:61]:
```suggestion
```py
from torch.utils.data import DataLoader

# Assuming custom_dataset is an instance of CustomDataset
data_loader = DataLoader(dataset=custom_dataset, batch_size=4, shuffle=True)

for batch in data_loader:
    samples, labels = batch
    # Training code here
```

```
Code should be written inside the ```py``` block and not the ```python``` block

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+## Built-in Datasets
+
+PyTorch also provides several built-in datasets that can be used with `DataLoader` without needing to define a custom dataset class. These are available in libraries such as `torchvision`, `torchaudio`, and `torchtext`.
+
+### Example: Loading a Built-in Dataset
+
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:68]:
```suggestion
### Example 


```
Instead of mentioning ` Loading a Built-in Dataset` in front of the example header, please note it in the description related to the example block.

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+```python
+from torchvision import datasets, transforms
+
+transform = transforms.Compose([
+    transforms.ToTensor(),
+    transforms.Normalize((0.5,), (0.5,))
+])
+
+train_dataset = datasets.FashionMNIST(root='data', train=True, download=True, transform=transform)
+train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:78]:
``` suggestion
```python
from torchvision import datasets, transforms

transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,))
])

train_dataset = datasets.FashionMNIST(root='data', train=True, download=True, transform=transform)
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)

```
Code should be written inside the ```py``` block and not the ```python``` block according to the [style-guide.md file](documentation/style-guide.md). Please have a look at it for a better understanding of how to write an entry in docs.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @wejk-ewjslkj,
I reviewed your PR and suggested a few changes. Please go through them asap and let me know once you go through the changes. Also, you can use the [documentation guide ](https://github.com/Codecademy/docs/tree/main/documentation)for better understanding and confusion if any.

Thanks,
Savi

**Comment by wejk-ewjslkj** [comment]
> Hey @wejk-ewjslkj, I reviewed your PR and suggested a few changes. Please go through them asap and let me know once you go through the changes. Also, you can use the [documentation guide ](https://github.com/Codecademy/docs/tree/main/documentation)for better understanding and confusion if any.
> 
> Thanks, Savi

I resolved all the issues you mentioned, except one. Pls have a look and let me know what you think. 
Cheers ..

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+---
+Title: 'Datasets and DataLoaders'
+Description: 'An overview of PyTorch Datasets and DataLoaders, including how to create custom datasets and use DataLoader for efficient data loading and batching.'
+Subjects: 
+  - 'Data Science'
+  - 'Machine Learning'
+Tags: 
+  - 'PyTorch'
+  - 'Datasets'
+  - 'DataLoader'
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:11]:
```suggestion
Tags: 
  - 'PyTorch'
  - 'Datasets'
  - 'DataLoader'
```
Please just include only those tags that are part of [tags.md](documentation/tags.md) list and you can add one if you want it to get included in the PR. So you can add the above tags to tags.md file.

File: content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+CatalogContent: 
+  - 'intro-to-pytorch-and-neural-networks'
+  - 'paths/build-a-machine-learning-model'
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/datasets-and-dataloaders/datasets-and-dataloaders.md:14]:
```suggestion
CatalogContent: 
  - 'intro-to-pytorch-and-neural-networks'
  - 'paths/build-a-machine-learning-model'
```
Please just include only those course/paths that are part of [catalog-content  list ](documentation/catalog-content.md)and you can add one if you want it to get included in the PR. So you can add the above paths in the catalog-content file.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @wejk-ewjslkj ,
I have suggested a few changes to you please make them asap, so I can approve it.

**Comment by SaviDahegaonkar** [comment]
Hey @wejk-ewjslkj ,
Have you implemented the changes I have suggested you, If yes then let me know so I can approve this PR.


Thanks,
Savi

**Review by SaviDahegaonkar — APPROVED** [review]
Hey @wejk-ewjslkj ,
Thank You for contributing to Codecademy Docs! The entry LGTM for Second round of review.

Thanks,
Savi

**Review by Sriparno08 — APPROVED** [review]
Looks good, @wejk-ewjslkj!
